internal/command/init.go

```diff
@@ -41,48 +43,57 @@ type InitCommand struct {
 }
 
 func (c *InitCommand) Run(args []string) int {
-	var flagFromModule, flagLockfile, testsDirectory string
-	var flagBackend, flagCloud, flagGet, flagUpgrade bool
 	var flagPluginPath FlagStringSlice
 	flagConfigExtra := newRawFlags("-backend-config")
 
+	var diags tfdiags.Diagnostics
 	args = c.Meta.process(args)
 	cmdFlags := c.Meta.extendedFlagSet("init")
-	cmdFlags.BoolVar(&flagBackend, "backend", true, "")
-	cmdFlags.BoolVar(&flagCloud, "cloud", true, "")
+	cmdFlags.Usage = func() {
+		diags = diags.Append(tfdiags.Sourceless(
+			tfdiags.Error,
+			"Failed to parse command-line flags",
+			c.Help(),
+		))
+	}
+
 	cmdFlags.Var(flagConfigExtra, "backend-config", "")
-	cmdFlags.StringVar(&flagFromModule, "from-module", "", "copy the source of the given module into the directory before init")
-	cmdFlags.BoolVar(&flagGet, "get", true, "")
-	cmdFlags.BoolVar(&c.forceInitCopy, "force-copy", false, "suppress prompts about copying state data")
-	cmdFlags.BoolVar(&c.Meta.stateLock, "lock", true, "lock state")
-	cmdFlags.DurationVar(&c.Meta.stateLockTimeout, "lock-timeout", 0, "lock timeout")
-	cmdFlags.BoolVar(&c.reconfigure, "reconfigure", false, "reconfigure")
-	cmdFlags.BoolVar(&c.migrateState, "migrate-state", false, "migrate state")
-	cmdFlags.BoolVar(&flagUpgrade, "upgrade", false, "")
 	cmdFlags.Var(&flagPluginPath, "plugin-dir", "plugin directory")
-	cmdFlags.StringVar(&flagLockfile, "lockfile", "", "Set a dependency lockfile mode")
-	cmdFlags.BoolVar(&c.Meta.ignoreRemoteVersion, "ignore-remote-version", false, "continue even if remote and local Terraform versions are incompatible")
-	cmdFlags.StringVar(&testsDirectory, "test-directory", "tests", "test-directory")
-	cmdFlags.Usage = func() { c.Ui.Error(c.Help()) }
-	if err := cmdFlags.Parse(args); err != nil {
+
+	initArgs, initDiags := arguments.ParseInit(args, cmdFlags)
```

Comment:
While attempting to do all the setup for init `cmdFlags` in the `arguments` package, I noticed that the init `cmdFlags` is tightly coupled to some constructs that are private to only the `command` package, examples: [newRawFlags](https://github.com/hashicorp/terraform/blob/d1761f436b636da959ad5865591d53edbe9df8b3/internal/command/init.go#L47) and [c.Meta.extendedFlagSet](https://github.com/hashicorp/terraform/blob/d1761f436b636da959ad5865591d53edbe9df8b3/internal/command/init.go#L50).

Possible solutions:
- For the `newRawFlags`, I see that a similar counterpart has been created [here](https://github.com/hashicorp/terraform/blob/main/internal/command/arguments/flags.go#L34) in arguments package, I can reuse that.

- For the `c.Meta.extendedFlagSet`, since the method mutates the Meta instance, we could pass the `c.Meta.extendedFlagSet` method as a value to `ParseInit` in arguments package or write a new `extendedFlagSet` in the argument package but pass in the required Meta instance values to `ParseInit`.

What do you think?

Reply:
Sounds reasonable!

The arguments package also has an `extendedFlagSet` implementation: https://github.com/hashicorp/terraform/blob/d1761f436b636da959ad5865591d53edbe9df8b3/internal/command/arguments/extended.go#L218-L258

Since we're already using the arguments package for `plan` and `apply`, I hope that should be sufficient for `init` too. If not, maybe we can improve it to be suitable.